### PR TITLE
Draft: Challenge X:  nexus credentials as part of a project pushed (#810)

### LINKED ISSUE #810: Nexus deployment credentials in settings.xml

### PR DESCRIPTION
**Pull request by divyanshuagarwal-23**
### What kind of changes does this PR include?
-Fixes

#### Description
As per #810 task one was to Create a settings.xml to connect to an imaginary Nexus repo, 
I have created the settings.xml file for the same
and have created a challenge reading the credential

### Relations
closes #810 

### References
issue: https://github.com/OWASP/wrongsecrets/issues/810
took ref from: https://github.com/sonatype/nexus-book-examples/blob/master/maven/settings/settings.xml

### CODE REVIEW

**Comment by divyanshuagarwal-23** [comment]
@commjoen please review this new PR, 
I will close the old one, which was causing issue due to challenge number

File: src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge42.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ */
+@Slf4j
+@Component
+@Order(42)
+public class Challenge42 extends Challenge {
+
+  private final Resource resource;
+
+  public Challenge42(
+      ScoreCard scoreCard, @Value("classpath:maven/settings/settings.xml") Resource resource) {
```
Comment (commjoen) [inline src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge42.java:31]: This resource does not exist (see below),  can you instead have a look at the code of Challenge12.java? here you can see how a file is loaded either from test resources or from a filepath within a docker container or from the resources.

File: src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge42.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  private String getSolution() {
+    try {
+      String config = resource.getContentAsString(Charset.defaultCharset());
+      StringReader stringReader = new StringReader(config);
+
+      XMLConfiguration xmlConfiguration = new XMLConfiguration();
+      xmlConfiguration.read(stringReader);
+
+      // Retrieve the Nexus password
+      return xmlConfiguration.getString("nexus.password");
```
Comment (commjoen) [inline src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge42.java:82]: this gives quite a few different exceptions. Can we do it with the jdk based parser?
```java
DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
        dbFactory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
        Document doc = dBuilder.parse(Paths.get(filepath "settings.xml").toFile());
        doc.getDocumentElement().normalize();
        return doc.getDocumentElement().getElementsByTagName("servers").item(0).getChildNodes().item(0).getNextSibling().getChildNodes().item(4).getNextSibling().getTextContent();
```
or something among these lines. And can you remove the entry from the pom.xml ?

File: src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge42.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    } catch (Exception e) {
+      log.warn("there was an exception with decrypting content in challenge42", e);
```
Comment (commjoen) [inline src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge42.java:84]:
```suggestion
      log.warn("there was an exception with retrieving and parsing the xml content for  challenge42", e);
```

File: src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge42.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+      return "error_decryption";
```
Comment (commjoen) [inline src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge42.java:85]:
```suggestion
      return "error_parsing";
```

File: src/test/java/org/owasp/wrongsecrets/challenges/docker/Challenge42Test.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    when(resource.getContentAsString(Charset.defaultCharset()))
+        .thenReturn(
+            "<root><nexus><username>test_user</username><password>test_password</password></nexus></root>");
+  }
+
+  @Test
+  void spoilerShouldGiveAnswer() {
+    var challenge = new Challenge42(scoreCard, resource);
+    Assertions.assertThat(challenge.spoiler().solution()).isNotEmpty();
+    Assertions.assertThat(challenge.answerCorrect(challenge.spoiler().solution())).isTrue();
```
Comment (commjoen) [inline src/test/java/org/owasp/wrongsecrets/challenges/docker/Challenge42Test.java:33]:
```suggestion
    Assertions.assertThat(challenge.answerCorrect(challenge.spoiler().solution())).isTrue();
     Assertions.assertThat(challenge.answerCorrect("error_parsing").isFalse();
```

**Review by commjoen — CHANGES_REQUESTED** [review]
I was not able to get it running on my system in this shape. Can you please have a look at the comments for an alternative approach?

File: src/main/resources/explanations/challenge42.adoc [hunk 6]
```diff
@@ -0,0 +1,3 @@
+=== Nexus credential read
+
+Storing nexus deployment credentials in your github project hardcoded is generally considered a bad practice because it undermines the security provided by encryption.
```
Comment (commjoen) [inline src/main/resources/explanations/challenge42.adoc:3]:
```suggestion
A developer wanted to configure his Nexus credentials and copy-pasted the configuration from the internal Confluence documentation site, while accidentally committing it to git!
Can you find the password?
```

File: src/main/resources/explanations/challenge42_reason.adoc [hunk 7]
```diff
@@ -0,0 +1,7 @@
+*Why you should not have nexus deployment credentials in your github project hardcoded*
+
+Storing nexus deployment credentials in your github project hardcoded is generally considered a bad practice because it undermines the security provided by encryption.
+
+In such scenarios, an attacker has the key the moment the file is in his possession.
+
+It is always recommended to store your credentials securely.
```
Comment (commjoen) [inline src/main/resources/explanations/challenge42_reason.adoc:7]:
```suggestion
Storing the credentials of your artifact registry in your Github project is a bad practice, as others can use these credentials to login, pull, and sometimes push artifacts in the name of the identity user of which the credentials were committed to git.

We often see the situation getting worse by just having one user hardcoded in many configurations of their projects.
Alternatively, we often see the artifact registry credentials hardcoded in the internal documentation site of an organization. 
In all cases, we often end up with a compromised user used to inject payloads into the artifact registry.

This is why it is better to have service accounts and/or personal accounts and/or a federated approach for users to the artifact registry. Next, make sure that you never hardcode the credentials: instead: have them configured as part of a secrets management setup.

```

**Comment by bendehaan** [comment]
Hello @divyanshuagarwal-23,

Thank you for your contribution. We appreciate your interest in helping improve WrongSecrets. However, we've noticed that there hasn't been any activity on this PR for several months.

We'll close this PR for now to maintain an efficient and organised workflow. Please don't be discouraged! We welcome you to reopen this PR or create a new one if you wish to continue your contribution in the future.

Feel free to reach out if you have any questions or need assistance.

Thank you for your understanding.

**Pull request closed without merging**